extensions/guacamole-auth-jdbc/modules/guacamole-auth-jdbc-base/src/main/java/org/apache/guacamole/auth/jdbc/tunnel/AccessEnforcingDelegatingTunnel.java

```diff
@@ -0,0 +1,196 @@
+     * wrapped tunnel.
+     */
+    private final AtomicReference<ModeledUser> user;
+
+    /**
+     * A thread that will continously refresh the user
+     */
+    private final Thread userRefreshThread;
+
+    /**
+     * A service to use for refreshing the user from the DB.
+     */
+    @Inject
+    private UserService userService;
+
+    /**
+     * Create a new tunnel that will enforce the access window restrictions of
+     * the provided user, during usage of the provided tunnel.
+     *
+     * @param tunnel
+     *     The tunnel to delegate to.
+     *
+     * @param modeledAuthenticatedUser
+     *     The user whose access restrictions should be applied.
+     *
+     */
+    @AssistedInject
+    public AccessEnforcingDelegatingTunnel(
+            @Nonnull @Assisted GuacamoleTunnel tunnel,
+            @Nonnull @Assisted ModeledAuthenticatedUser modeledAuthenticatedUser) {
+
+        super(tunnel);
+        this.user = new AtomicReference<>(modeledAuthenticatedUser.getUser());
+
+        this.userRefreshThread = new Thread(() -> {
+            while (true) {
+
+                try {
+
+                    // Fetch an up-to-date user record from the DB to ensure
+                    // that any access restrictions modified while this tunnel
+                    // is open will be taken into account
+                    this.user.set(userService.retrieveUser(
+                        modeledAuthenticatedUser.getAuthenticationProvider(),
+                        modeledAuthenticatedUser));
+                }
+
+                // If an error occurs while trying to fetch the updated user,
+                // log the warning / exception and stop the refresh thread
+                catch (GuacamoleException e) {
+
+                    logger.warn(
+                            "Aborting user refresh thread due to error: {}",
+                            e.getMessage());
+                    logger.debug(
+                            "Exception caught while attempting to refresh user.", e);
+
+                    return;
+                }
+
+                try {
+
+                    // Wait a bit before refreshing the user record again
+                    Thread.sleep(USER_MODEL_REFRESH_INTERVAL);
+                }
+
+                // If interrupted by the tunnel, exit immediately
+                catch (InterruptedException e) {
+                    return;
+                }
+            }
+        });
+    }
+
+    @Override
+    public GuacamoleReader acquireReader() {
+
+        // Start periodically refreshing the user record
+        userRefreshThread.start();
+
+        // Filter received instructions, checking if the user's login
+        // is still valid for each one. If the login is invalid,
+        // log them out immediately and close the tunnel.
+        return new FilteredGuacamoleReader(
+            super.acquireReader(),
+            new GuacamoleFilter() {
```

Comment:
Rather than override `acquireReader()`, etc., it would be better to override `getSocket()` and use a `FilteredGuacamoleSocket`. The same filter implementation could be used for both reads and writes.